# Practice Case 01

## Initial Data Analysis of German Credit Score Dataset

In [46]:
import pandas as pd
german = pd.read_csv('D:\Data Science Bootcamp\Day 4\german_credit_data_with_risk.csv', encoding='latin1')

In [36]:
german.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 11 columns):
Unnamed: 0          1000 non-null int64
Age                 1000 non-null int64
Sex                 1000 non-null object
Job                 1000 non-null int64
Housing             1000 non-null object
Saving accounts     817 non-null object
Checking account    606 non-null object
Credit amount       1000 non-null int64
Duration            1000 non-null int64
Purpose             1000 non-null object
Risk                1000 non-null object
dtypes: int64(5), object(6)
memory usage: 86.1+ KB


# Part A
On your Jupyter notebook, answer the following questions! 


1. How many rows and columns reflected in the dataset?
<br> *__There are 1000 rows (excluding header) and 11 columns (including index) in the dataset.__*

In [37]:
german.shape

(1000, 11)

2. What is the minimum, maximum, and mean of the customer's age?
<br> *__From the describe method below:__*
<br>    *__- The minimum of customer's age: 19__*
<br>    *__- The maximum of customer's age: 75__*
<br>    *__- The mean of customer's age: 35.546__*

In [38]:
german.describe()

Unnamed: 0          Age          Job  Credit amount     Duration
count  1000.000000  1000.000000  1000.000000    1000.000000  1000.000000
mean    499.500000    35.546000     1.904000    3271.258000    20.903000
std     288.819436    11.375469     0.653614    2822.736876    12.058814
min       0.000000    19.000000     0.000000     250.000000     4.000000
25%     249.750000    27.000000     2.000000    1365.500000    12.000000
50%     499.500000    33.000000     2.000000    2319.500000    18.000000
75%     749.250000    42.000000     2.000000    3972.250000    24.000000
max     999.000000    75.000000     3.000000   18424.000000    72.000000

3. How many are male and how many are female? Why?
<br> *__There are 690 males and 310 females.__*
<br> *__It is due to the majority of males do not have checking accounts with percentage of 27.8% and also in terms of saving accounts, there are about 40.9% of males who have little saving accounts, compared to females at 19.4%. Also, there are also more males who do not have saving accounts (13.4%) compared to females. There might be another reason on this. Since males are more into automotives, according to the dataset, there are more males who have purpose in purchasing cars at 24.3% compared to females at 9.4%.__*

In [39]:
german['Sex'].value_counts()

male      690
female    310
Name: Sex, dtype: int64

4. How many high skilled workers customers reflected in the dataset?
<br> *__There are 630 skilled-worker customers in the dataset, which is the majority of the job employment type.__*

In [40]:
german['Job'].value_counts()

2    630
1    200
3    148
0     22
Name: Job, dtype: int64

5. How many customers who own the house?
<br> *__There are 713 customers who own the house.__*

In [41]:
german['Housing'].value_counts()

own     713
rent    179
free    108
Name: Housing, dtype: int64

6. How many customers who have 'rich' Saving accounts?
<br> *__There are 48 customers who have 'rich' saving accounts, which is the minority of the dataset.__*

In [42]:
german['Saving accounts'].value_counts()

little        603
moderate      103
quite rich     63
rich           48
Name: Saving accounts, dtype: int64

7. How many customers who do not have checking accounts or have zero checking account?
<br> *__There are 394 customers who do not have checking accounts, which is the majority of the dataset.__*

In [43]:
german['Checking account'].value_counts()

little      274
moderate    269
rich         63
Name: Checking account, dtype: int64

In [94]:
german['Checking account'].fillna('N/A', inplace = True)  

In [95]:
german['Checking account'].value_counts()

N/A         394
little      274
moderate    269
rich         63
Name: Checking account, dtype: int64

8. Who is the one who have highest credit amount? Show us the profile of the age, sex, job, housing, saving account, and checking account!
<br> *__Customer who has the highest credit amount is a middle-aged woman (32 years old) who is a high-skilled worker and owns a house. She has little saving account and also moderate checking account.__*

In [56]:
german[['Age','Sex','Job','Housing','Saving accounts','Checking account','Credit amount']].sort_values(by=['Credit amount'], ascending=False)[:1]

Age     Sex  Job Housing Saving accounts Checking account  Credit amount
915   32  female    3     own          little         moderate          18424

9. How many customers are classified as good risk and bad risk? 
<br> *__There are 700 customers classified as good risk and also 300 customers classified as bad risk.__*

In [57]:
german['Risk'].value_counts()

good    700
bad     300
Name: Risk, dtype: int64

10. Is the average of credit amount for good risk is higher than bad risk?
<br> *__No. The average of credit amount for good risk is lower than bad risk.__*

In [66]:
german[german['Risk']=='good']['Credit amount'].mean()

2985.457142857143

In [67]:
german[german['Risk']=='bad']['Credit amount'].mean()

3938.1266666666666

# Part B

Aside questions mentioned on Part A, create 10 more questions and answers to understand more about the dataset. 


1. What is the highest credit amount for males whose house is free and is skilled?
<br> *__The highest credit amount for males whose house is free and is skilled is 12,612 DM.__*

In [74]:
german[(german['Sex']=='male') & (german['Housing']=='free') & (german['Job']==2)][['Credit amount']].sort_values(by=['Credit amount'], ascending = False)[:1].reset_index(drop=True)

Credit amount
0          12612

2. Which duration (in months) that is the most popular for Germans?
<br> *__24 months is the majority for the installment option.__*

In [75]:
german['Duration'].value_counts()

24    184
12    179
18    113
36     83
6      75
15     64
9      49
48     48
30     40
21     30
10     28
27     13
60     13
42     11
11      9
20      8
8       7
4       6
39      5
45      5
7       5
14      4
13      4
33      3
28      3
22      2
16      2
54      2
26      1
40      1
47      1
5       1
72      1
Name: Duration, dtype: int64

3. What purpose that is the majority for both males and females?
<br> *__Car is the major purpose for both males and females.__*

In [78]:
german['Purpose'][german['Sex']=='male'].value_counts()

car                    243
radio/TV               195
furniture/equipment    107
business                78
education               35
repairs                 17
vacation/others          9
domestic appliances      6
Name: Purpose, dtype: int64

In [79]:
german['Purpose'][german['Sex']=='female'].value_counts()

car                    94
radio/TV               85
furniture/equipment    74
education              24
business               19
domestic appliances     6
repairs                 5
vacation/others         3
Name: Purpose, dtype: int64

4. How many customers who rent a house but have so much savings?
<br> *__There are 9 customers who rent a house but have so much savings.__*

In [87]:
german[(german['Housing']=='rent') & (german['Saving accounts']=='rich')]['Age'].count()

9

5. What is the average age for females who owns a house?
<br> *__The average age for females who owns a house is 33.67.__*

In [90]:
german[(german['Sex']=='female') & (german['Housing']=='own')]['Age'].mean()

33.66836734693877

6. How many customers who own a house but have no checking account?
<br> *__There are 304 customers who own a house, but have no checking accounts.__*

In [100]:
german[(german['Housing']=='own') & (german['Checking account']=='N/A')]['Housing'].count()

304

7. For customers who purchased auto loans, what is the average tenor and what is the max tenor?
<br> *__The average tenor who purchased auto loans is 20.8 months.__*
<br> *__The max tenor who purchased auto loans is 60 months.__*

In [103]:
german[german['Purpose']=='car'][['Duration']].mean()

Duration    20.813056
dtype: float64

In [104]:
german[german['Purpose']=='car'][['Duration']].max()

Duration    60
dtype: int64

8. What is the lowest credit amount for customer whose purpose is car for both males and females?
<br> *__The lowest credit amount for male customers whose purpose is car is 276 DM.__*
<br> *__The lowest credit amount for female customers whose purpose is car is 250 DM.__*

In [110]:
german[(german['Purpose']=='car') & (german['Sex']=='male')][['Credit amount']].sort_values(by=['Credit amount'], ascending = True)[:1]

Credit amount
309            276

In [111]:
german[(german['Purpose']=='car') & (german['Sex']=='female')][['Credit amount']].sort_values(by=['Credit amount'], ascending = True)[:1]

Credit amount
725            250

9. How many percentage whose risk is good and bad for those who have credits in cars?
<br> *__There are 23.1% of customers who have credits in cars are classified as good risk and 10.6% of customers who have credits in cars are classified as bad risk.__*

In [129]:
count_good = len(german['Risk'][(german.Risk == 'good') & (german.Purpose == 'car')]) 
count_bad = len(german['Risk'][(german.Risk == 'bad') & (german.Purpose == 'car')])
count_risk = len(german['Risk'])
percent_good = round((count_good/count_risk)*100,2)
percent_bad = round(((count_bad/count_risk)*100),2)

In [132]:
print(str(percent_good) + '%')
print(str(percent_bad) + '%')

23.1%
10.6%


10. What is the majority of job type for good risk and bad risk? What is the percentage for each classification?
<br> *__The majority of job type for good risk and bad risk is Job Type 2, which is skilled-worker.__*
<br> *__The percentage of both good risk and bad risk is 44.6% and 18.6%.__*

In [141]:
# First is to find out the highest quantity on job type for good risk
german[german['Risk']=='good']['Job'].value_counts()

2    444
1    144
3     97
0     15
Name: Job, dtype: int64

In [142]:
# Now for the bad risk
german[german['Risk']=='bad']['Job'].value_counts()

2    186
1     56
3     51
0      7
Name: Job, dtype: int64

In [143]:
# After that find the percentage for each class
count_good = len(german['Risk'][(german.Risk == 'good') & (german.Job == 2)]) 
count_bad = len(german['Risk'][(german.Risk == 'bad') & (german.Job == 2)])
count_risk = len(german['Risk'])
percent_good = round((count_good/count_risk)*100,2)
percent_bad = round(((count_bad/count_risk)*100),2)

In [134]:
print(str(percent_good) + '%')
print(str(percent_bad) + '%')

44.4%
18.6%


# Part C
Define at least 3 business problems that you can find based on the dataset! 


*__- Since we all know that the percentage for good risk is 70% and bad risk is 30%, business must need to cut loss even more. In this case, business want to come up with new ideas/programs to decrease the bad risk percentage by 20% or even more.__*
<br>
*__- To better analyze the data and also build more accurate models for the data, business will need to do customer data update, especially in the checking account classification, since there are a lot of null values in checking account variable.__* 
<br>
*__- The current dataset might not be reliable just to judge based on the risk itself. To come up with the risk category, the data will need score category of each customer to judge if the customer is good or not. Then, make a score cut-off to decide if customer is in good risk or bad risk.__* 